src/main/java/htsjdk/samtools/cram/compression/nametokenisation/NameTokenisationDecode.java

```diff
@@ -0,0 +1,164 @@
+                    currentToken = leftPadNumber(delta0Token, lenDelta0Token);
+                    break;
+                case TOKEN_MATCH:
+                    currentToken = tokensList.get(prevNameIndex).get(tokenPosition-1);
+                    break;
+                default:
+                    break;
+            }
+            tokensList.get(currentNameIndex).add(tokenPosition-1,currentToken);
+            decodedNameBuilder.append(currentToken);
+            tokenPosition++;
+        } while (type!= TOKEN_END);
+        return decodedNameBuilder.toString();
+        }
+
+    private static String getDeltaToken(
+            final TokenStreams tokenStreams,
+            final int tokenPosition,
+            final List<List<String>> tokensList,
+            final int prevNameIndex,
+            final byte tokenType) {
+        if (!(tokenType == TOKEN_DELTA || tokenType == TOKEN_DELTA0)){
+            throw new CRAMException(String.format("Invalid tokenType : %s. " +
+                    "tokenType must be either TOKEN_DELTA or TOKEN_DELTA0", tokenType));
+        }
+        int prevToken;
+        try {
+            prevToken = Integer.parseInt(tokensList.get(prevNameIndex).get(tokenPosition -1));
+        } catch (final NumberFormatException e) {
+            final String exceptionMessageSubstring = (tokenType == TOKEN_DELTA) ? "DIGITS or DELTA" : "DIGITS0 or DELTA0";
+            throw new CRAMException(String.format("The token in the prior name must be of type %s",
+                    exceptionMessageSubstring), e);
+        }
+        final int deltaTokenValue = tokenStreams.getTokenStreamByteBuffer(tokenPosition,tokenType).get() & 0xFF;
+        return Long.toString(prevToken + deltaTokenValue);
+    }
+
+    private static String getDigitsToken(
+            final TokenStreams tokenStreams,
+            final int tokenPosition,
+            final byte tokenType ) {
+        if (!(tokenType == TOKEN_DIGITS || tokenType == TOKEN_DIGITS0)){
+            throw new CRAMException(String.format("Invalid tokenType : %s. " +
+                    "tokenType must be either TOKEN_DIGITS or TOKEN_DIGITS0", tokenType));
+        }
+        final ByteBuffer digitsByteBuffer = tokenStreams.getTokenStreamByteBuffer(tokenPosition, tokenType).order(ByteOrder.LITTLE_ENDIAN);
+        final long digits = digitsByteBuffer.getInt() & 0xFFFFFFFFL;
+        return Long.toString(digits);
+    }
+
+    private static String readString(final ByteBuffer inputBuffer) {
+        // spec: We fetch one byte at a time from the value byte stream,
+        // appending to the name buffer until the byte retrieved is zero.
+        StringBuilder resultStringBuilder = new StringBuilder();
+        byte currentByte = inputBuffer.get();
+        while (currentByte != 0) {
+            resultStringBuilder.append((char) currentByte);
+            currentByte = inputBuffer.get();
+        }
```

Comment:
Not a big deal in this case, but in the spirit of keeping narrow variable scope, I would always prefer something like this:
```suggestion
        for (byte currentByte = inputBuffer.get(); currentByte != 0; currentByte = inputBuffer.get()) {
            resultStringBuilder.append((char) currentByte);
        }
```